src/sync/sync_manager.hpp

```diff
@@ -207,10 +217,10 @@ friend class SyncSession;
     bool run_file_action(const SyncFileActionMetadata&);
 
     // Protects m_users
-    mutable std::mutex m_user_mutex;
+    mutable std::recursive_mutex m_user_mutex;
```

Comment:
Rather than changing this to a recursive_mutex (which makes it pretty easy to write subtle deadlocks as most of the things it enables are inherently unsafe), log_out_user() could just call m_metadata_manager->set_current_user_identity() directly.